sips/sip-45.md

```diff
@@ -0,0 +1,78 @@
+
+We have located and analyzed multiple CEX-DEX trading bots and found no exception other than using a constant gas price (for example, 850 MIST), despite the fact that they are competing with each other.
+
+As of today, there is little one can do about the case, other than submitting many different transactions that are further submitted by different validators, or simply resort to luck.
+
+We believe one way to resolve this is to enable the user to send a tx with a even higher gas price (that comes with a significant real cost) should they choose to do so, while if the price paid is high enough, more validators will submit the transaction immediately, referencing its computed submission position from transaction’s digest. The higher gas price the user pays, the more validators will submit the transaction immediately. We have chosen the function of N so that a user has no incentive to send multiple transactions.
+
+We have studied transactions in the past 30 days (as of 2024-11-07), a distribution of gas price as well as the projected effects from our proposal are shown as follows:
+
+| Gas Price Bucket | Number of Txs | %      | $max(N)$ | `Amp%`   | `Cum Amp%` |
+| ---------------: | :------------ | ------ | ------ | ------ | -------- |
+| $[0, 5000)$        | 249058741     | 99.783 | 1      | 99.783 | 99.783   |
+| $[5000, 10000)$    | 37872         | 0.015  | 14     | 0.212  | 99.996   |
+| $[10000, 20000)$   | 303449        | 0.122  | 27     | 3.283  | 103.278  |
+| $[20000, 30000)$   | 58921         | 0.024  | 40     | 0.944  | 104.222  |
+| $[30000, 40000)$   | 5370          | 0.002  | 54     | 0.116  | 104.339  |
+| $[40000, 50000)$   | 4128          | 0.002  | 67     | 0.111  | 104.449  |
+| $[50000, 60000)$   | 9958          | 0.004  | 80     | 0.319  | 104.769  |
+| $[60000, 70000)$   | 2818          | 0.001  | 94     | 0.106  | 104.875  |
+| $[70000, 80000)$   | 3953          | 0.002  | 107    | 0.169  | 105.044  |
+| $[80000, 90000)$   | 1729          | 0.001  | 120    | 0.083  | 105.127  |
+| $[90000, 100000)$  | 112992        | 0.045  | 134    | 6.066  | 111.193  |
+
+- `Amp%` represents the percentage of amplification from that bucket.
+- `Cum Amp%` represents the cumulative % of amplification from that bucket and all preceding buckets.
+
+From the data above, we predict that consensus bandwidth will be increased by no more than 12% should the proposal get implemented. In return, non-deterministic factors will become minimum, which makes actors able to compete by trying to outbid each other, where the participants will have to pay a much higher gas fee than before, in exchange for the certainty of reliably winning, should one outbid everyone else.
```

Comment:
Unclear why consensus bandwidth is expected to increase by 12%. Removing latency jitters could reduce effective throughput by allowing more duplicates. Also, high-priority transactions may increase object dependencies, reducing execution throughput and, by feedback, consensus throughput. While this issue likely existed before, the proposed changes make users pay more.

Reply:
We assumed most on-chain txs are not competitive in nature (meaning that they are not competing to land before the others do).  For those txs, having priority does not change anything.

For competitive txs, once there is priority, the number of txs will be reduced (actors will seek paying higher prices rather than sending more, because simply sending more does not work anymore) - this may slightly increase consensus bandwidth required.

The proposal naturally works with the object congestion, since the congestion control algorithm prefers to select txs with higher gas prices.  When there is a congestion (this often but not always means there is a competition), txs with higher gas price win.  In this sense, high-priority txs do not lead to congestion, market does.  For users, this brings more fairness - whoever pays more gets sequenced first, if there is a competition.

Reply:
Iiuc the increase is in effective throughput, it is expected to come from fewer transaction duplicates for these high-priority transactions.  If I am willing to pay a high price, why should I not send my transaction to more than 1 validator, say K? It would increase my chances of getting into the current checkpoint. Then, in the worst case, we have K copies of my transaction, possibly decreasing the effective throughput.

Reply:
Yes.  With higher gas prices, `N` gets bigger, meaning more 'copies' of the transaction will be sent to consensus.

The term 'consensus bandwidth' from the SIP actually refers to network bandwidth (or put it another way, network traffic), not the capacity for consensus to process txs.

The data is to show that despite more copies will be sent over the network, the percentage of competitive txs is small, which do not lead to a significant increase overall.